hop/io.py

```diff
@@ -214,74 +211,129 @@ def _generate_group_id(n):
 
 
 @dataclass
-class _Metadata:
+class Metadata:
     """Broker-specific metadata that accompanies a consumed message.
 
     """
 
-    topic: str
-    partition: int
-    offset: int
-    timestamp: int
-    key: Union[str, bytes]
+    topic: str = field(init=False)
+    partition: int = field(init=False)
+    offset: int = field(init=False)
+    timestamp: int = field(init=False)
+    key: Union[str, bytes] = field(init=False)
+    _raw: confluent_kafka.Message
+
+    def __post_init__(self):
+        self.topic = self._raw.topic()
+        self.partition = self._raw.partition()
+        self.offset = self._raw.offset()
+        self.timestamp = self._raw.timestamp()[1]
+        self.key = self._raw.key()
```

Comment:
This works, of course, but I would have expected a classmethod constructor I think:
```python
@classmethod
def from_message(cls, msg: confluent_kafka.Message) -> Metadata:
    return Metadata(
        topic=msg.topic(),
        partition=msg.partition(),
        offset=msg.offset(),
        timestamp=msg.timestamp()[1],
        key=msg.key(),
        _raw=msg,
    )
```

Obviously, just a matter of taste, but I like that you don't need the noisy `field(init=False)` annotations, and constructing things as `Metadata.from_message(msg)` seems tidy.

Reply:
Yup, I totally agree with that, it looks much nicer. I had to change the return annotation to the `'Metadata'` string to get this to work, otherwise get a `NameError: name 'Metadata' is not defined` error. Have you encountered that issue when adding type annotations to adc-streaming?

Reply:
Ah yep. The reason is that there's no forward definitions in type annotations - the python interpreter walks through the file in one pass, so it can't refer to things that come later in the file, or do anything self-referential.

[PEP-563 -- Postponed Evaluation of Annotations](https://www.python.org/dev/peps/pep-0563/) proposes ways to solve this problem. In Python 3.7+, you can use `from __future__ import annotations`. In Python 3.10, it'll just work. In Python 3.6-, you're out of luck, and simply must use strings.